### ISSUE TITLE
[Bug]: Table multi-selection always selects the next item by default when an item above it is been removed

### ISSUE THREAD

**Issue by wendyyuanxie** [microsoft/fluentui#28456]
### Library

React Components / v9 (@fluentui/react-components)

### System Info

```shell
System:
    OS: Windows 10 10.0.22621
    CPU: (20) x64 12th Gen Intel(R) Core(TM) i7-1280P
    Memory: 14.52 GB / 31.67 GB
  Browsers:
    Edge: Spartan (44.22621.1848.0), Chromium (114.0.1823.67)
    Internet Explorer: 11.0.22621.1
```

### Are you reporting Accessibility issue?

None

### Reproduction

https://codesandbox.io/s/nameless-monad-mhhn5j?file=/example.tsx

## Actual Behavior
This behavior can be found for multi-selection of table component in flunetV9.
I am able to reproduce it in the codeSandbox by using one of the official example: https://codesandbox.io/s/nameless-monad-mhhn5j?file=/example.tsx
Let's say we have 4 items in the table like this: 
<img width="517" alt="image" src="https://github.com/microsoft/fluentui/assets/125315752/c5db76c3-2925-4d26-8b55-fcda5cfae6bd">

and we remove the first item from it by clicking on the delete:
<img width="522" alt="image" src="https://github.com/microsoft/fluentui/assets/125315752/c91e5d41-56dd-4460-aff3-5b2ae412ed67">

After the selected item is been removed, the next item is automatically get selected be default.

## Expected Behavior
After the selected item is been removed, the next item should remain unselected state

### Requested priority

High

### Are you willing to submit a PR to fix?

no

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

**Comment by behowell** [comment]
Hi @wendyyuanxie, it looks like your repro example is using the index of the row to track selection (the default value of `rowId` is the row's index). That's causing problems when you delete an item,, which changes the index of the items.

Instead, you can use another key as the row ID, and pass a custom `getRowId` to the `useTableFeatures` function. 

See https://codesandbox.io/s/mutable-forest-3l66hz?file=/example.tsx for an update to your repro example, which uses the `item.file.label` as the row ID.

Please let me know if this doesn't solve the issue. Thanks.

**Comment by wendyyuanxie** [comment]
@behowell Thanks for the example and explanations. But I don't think it fully resolves the issue. For example, if I selected the second one, and deleted it:
![image](https://github.com/microsoft/fluentui/assets/125315752/3fcac24c-6f4a-423c-8d79-3fb6e9d2705e)
and it will become this: 
![image](https://github.com/microsoft/fluentui/assets/125315752/631cb138-8cbc-4fe2-b89d-459bc324fec6)


Also, for some reason, when we remove an item from the list, the checkbox at the column names level will always remain selected: 
![image](https://github.com/microsoft/fluentui/assets/125315752/a0d173a1-571e-4c4a-a3f0-9865151d7cbe)

**Comment by ling1726** [comment]
Thanks for reporting @wendyyuanxie. as @behowell  mentioned, in these cases you should explicitly set some kind of identifier for the row and use the `getRowId` prop. The `allRowsSelected` and `someRowsSelected` states aren't as reliable as they should be and will be fixed in #28577.

Currently as a workaround you can:
* set/find an identifier for each row of your data
* use getRowId
* call `deselectRow` after you delete a row(s)

here's an example of the workaround https://codesandbox.io/s/dreamy-blackburn-dwlyn8?file=/example.tsx
